### PR TITLE
[WIP] Use react submit button in forms

### PR DESCRIPTION
**Pull request by boaz0**
I took an initiative and replaced the old submit button with a `react` one.
I would like to hear your opinion and get your feedback.

I haven't created a ticket in Redmine because I wasn't sure what you would think.

@xprazak2 @sharvit @amirfefer @ohadlevy @gilad215

Thanks

### CODE REVIEW

File: webpack/assets/javascripts/react_app/components/SubmitButton/index.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    const { data: props } = this.props;
+    return (
+      <button
+        disabled={this.state.disabled}
+        onClick={() => {
+          this.setState({ disabled: !this.state.disabled });
+          return true;
+        }}
+        {...props}>
+        Submit
```
Comment (waldenraines) [inline webpack/assets/javascripts/react_app/components/SubmitButton/index.js]: This isn't translated, needs `{__('Submit')}`

**Review by waldenraines — CHANGES_REQUESTED** [review]

**Comment by ohadlevy** [comment]
Can we reuse the existing (react) form submit?

Also, does this makes sense as a standalone component?

**Comment by boaz0** [comment]
@ohadlevy thank you for letting me know about `Actions.js`. I am using this and the next step is probably to replace the current class `SubmitCancelButtons` with `redux`.

Nevertheless, I am wondering why did we use `data-id` in the `Submit` button?
Does it have any important role that can't be implemented in `react`?

Thanks.

File: webpack/assets/javascripts/react_app/redux/actions/user/submitCancelButtons/index.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+});
+
+export const onCancelClick = () => ({
+  type: CANCEL_BUTTON_CLICKED,
+  payload: true,
+});
+
+export const onMountButtons = () => ({
+  type: RESET_SUBMIT_AND_CANCEL,
+  payload: false,
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/redux/actions/user/submitCancelButtons/index.js]: Why do they need payloads?
Isn't it enough:
```js
export const onMountButtons = () => ({ type: RESET_SUBMIT_AND_CANCEL });
```

File: webpack/assets/javascripts/react_app/components/SubmitCancelButtons/index.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+          windowLocation.replace(cancelPath);
+        }}
+        onSubmitClick={submitDispatch}
+      />
+    );
+  }
+}
+
+const mapStateToProps = (state) => {
+  const { submitCancelButtons } = state;
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/SubmitCancelButtons/index.js]: Will save a line :grin: 
```js
const mapStateToProps = ({ submitCancelButtons }) => { ... };
```

File: webpack/assets/javascripts/react_app/components/SubmitCancelButtons/index.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  return {
+    submitting: submitCancelButtons.submitting || false,
+    disabled: submitCancelButtons.disabled || false,
+    windowLocation: submitCancelButtons.windowLocation || window.location,
+  };
+};
+
+const mapDispatchToProps = dispatch =>
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/SubmitCancelButtons/index.js]: Check this out, a simpler way to attach action-dispatchers:
https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/BreadcrumbBar/index.js#L21

File: webpack/assets/javascripts/react_app/components/SubmitCancelButtons/SubmitCancelButtons.test.js [hunk 5]
```diff
@@ -0,0 +1,203 @@
+import toJson from 'enzyme-to-json';
+import { shallow } from 'enzyme';
+import React from 'react';
+import { SubmitCancelContainer } from './index';
+import Actions from '../common/forms/Actions';
+
+const data = { cancelPath: 'test' };
+const onMountButtons = jest.fn();
+
+describe('submit_or_cancel_buttons', () => {
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/SubmitCancelButtons/SubmitCancelButtons.test.js]: We found out it is much easier to test **unconnected** components against snapshots.
considre exporting both connected and unconnected to make it more reusable and testable.

**Review by sharvit — COMMENTED** [review]
Thanks @ripcurld :+1: 
I like the way the code looks. Waiting to see the [WIP] getting removed.

File: webpack/assets/javascripts/react_app/common/testHelpers.js [hunk 6]
```diff
@@ -56,3 +56,14 @@ export const testComponentSnapshotsWithFixtures = (Component, fixtures) =>
   shallowRenderComponentWithFixtures(Component, fixtures)
     .forEach(({ description, component }) =>
       it(description, () => expect(toJson(component)).toMatchSnapshot()));
+
+/**
+ * Test a reducer with fixtures and snapshots
+ * @param  {Function} reducer  reduce function to test
+ * @param  {Object}   fixtures key=fixture description, value=props to apply
+ */
+export const testReducerSnapshotWithFixtures = (reducer, fixtures) => {
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/common/testHelpers.js:65]: Awesome :+1:

Comment (sharvit) [reply webpack/assets/javascripts/react_app/common/testHelpers.js:65]: Can you actually separate it (and the consuming of that at `BreadcrumbBarReducer.test.js`) into a different pr?

I want to start using it and I think we can merge it quickly.

Comment (boaz0) [reply webpack/assets/javascripts/react_app/common/testHelpers.js:65]: Yeah sure, I will work on it today.

Comment (boaz0) [reply webpack/assets/javascripts/react_app/common/testHelpers.js:65]: @sharvit opened https://github.com/theforeman/foreman/pull/5516 to do as you asked

**Comment by ohadlevy** [comment]
@ripcurld I'm a bit confused to why this PR make sense :) what is the motivation to generate just the submit/cancel buttons on current html forms ?

**Comment by boaz0** [comment]
@ohadlevy that's a good question and it's the main reason I didn't remove "WIP" from the title.
I guess I'll close it till we will start moving the forms into React components.

**Pull request closed without merging**